completable_html_xblock/html.py

```diff
@@ -0,0 +1,68 @@
+"""This XBlock help creating a secure and easy-to-use HTML blocks in edx-platform."""
+from __future__ import absolute_import
+
+import logging
+
+import pkg_resources
+from html_xblock import HTML5XBlock
+from html_xblock.utils import _  # pylint: disable=protected-access
+from xblock.completable import CompletableXBlockMixin
+from xblock.core import XBlock
+from xblock.fields import Scope, String
+from xblockutils.resources import ResourceLoader
+
+log = logging.getLogger(__name__)  # pylint: disable=invalid-name
+xblock_loader = ResourceLoader(__name__)  # pylint: disable=invalid-name
+
+
+class CompletableHTML5XBlock(CompletableXBlockMixin, HTML5XBlock):
+    """
+    This XBlock will disable completion and add provide and  an HTML WYSIWYG interface in Studio to be rendered in LMS.
+    """
+
+    display_name = String(
+        display_name=_('Display Name'),
+        help=_('The display name for this component.'),
+        scope=Scope.settings,
+        default=_('Completable')
+    )
+    allow_javascript = True
+    editor = String(
+        help=_(
+            'Select Visual to enter content and have the editor automatically create the HTML. Select Raw to edit '
+            'HTML directly. If you change this setting, you must save the component and then re-open it for editing.'
+        ),
+        display_name=_('Editor'),
+        default='raw',
+        values=[
+            {'display_name': _('Visual'), 'value': 'visual'},
+            {'display_name': _('Raw'), 'value': 'raw'}
+        ],
+        scope=Scope.settings
+    )
+    has_custom_completion = True
+    editable_fields = ('display_name', 'editor')
+
+    @staticmethod
+    def completable_resource_string(path):
+        """We need to subclass this, because we don't want to override its usages in superclass methods."""
+        data = pkg_resources.resource_string(__name__, path)
+        return data.decode('utf8')
+
+    @XBlock.supports('multi_device')
+    def student_view(self, context=None):  # pylint: disable=unused-argument
+        """
+        Return a fragment that contains the html for the student view. Add #complete element.
+        """
+        frag = super(CompletableHTML5XBlock, self).student_view()
+        frag.add_javascript(self.completable_resource_string('static/js/html_completion.js'))
+        frag.initialize_js('HTML5CompletionXBlock')
+
+        return frag
+
+    @XBlock.json_handler
+    def complete(self, _data, _suffix=''):
+        """
+        Use new completion API for marking the block as completed.
+        """
+        self.emit_completion(1.0)
```

Comment:
FYI there is already a default `json_handler` called `publish_completion` that gets automatically mixed in to every XBlock in the LMS: https://github.com/edx/edx-platform/blob/1bcaa945d9ba9ffb4cc14131d782123fb9e62a94/lms/djangoapps/lms_xblock/mixin.py#L250-L263

So you probably don't need this handler, you an just call that existing one.

Reply:
It didn't seem to be working for me, because [this check](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/lms_xblock/mixin.py#L260-L261) was failing (as [this](https://github.com/edx/completion/blob/master/completion/services.py#L110-L119) was returning `False`) when I tried using it with `has_custom_completion = True` (to disable automatic completion on view). I might've omitted something here, though.

Reply:
Ah I see. Based on that check if you se `has_custom_completion = True` it won't allow you to use that handler.